lib/openstax/aws/stack.rb

```diff
@@ -218,14 +222,80 @@ def apply_change_set(params: {}, wait: false)
           logger.info("Executing change set")
           change_set.execute
           reset_cached_remote_state
-        end
 
-        wait_for_update if wait
+          if wait
+            wait_for_update
+            tag_alarms_and_rules
+          end
+        end
+      else
+        tag_alarms_and_rules
       end
 
       change_set
     end
 
+    def autoscaling_client
+      @autoscaling_client ||= Aws::AutoScaling::Client.new region: region
+    end
+
+    def cloudwatch_client
+      @cloudwatch_client ||= Aws::CloudWatch::Client.new region: region
+    end
+
+    def eventbridge_client
+      @eventbridge_client ||= Aws::EventBridge::Client.new region: region
+    end
+
+    def tag_alarms_and_rules
+      account_id = Aws::STS::Client.new(region: region).get_caller_identity.account
+      stack_tags = self.class.format_hash_as_tag_parameters @tags
+
+      client.list_stack_resources(stack_name: name).each do |response|
+        response.stack_resource_summaries.each do |stack_resource_summary|
+          resource_id = stack_resource_summary.physical_resource_id
+
+          case stack_resource_summary.resource_type
+          when 'AWS::CloudWatch::Alarm'
+            resource_arns = [ "arn:aws:cloudwatch:#{region}:#{account_id}:alarm:#{resource_id}" ]
+            resource_client = cloudwatch_client
+
+          when 'AWS::Events::Rule'
+            resource_arns = [ "arn:aws:events:#{region}:#{account_id}:rule/#{resource_id}" ]
+            resource_client = eventbridge_client
+          when 'AWS::AutoScaling::AutoScalingGroup'
+            resource_arns = autoscaling_client.describe_policies(
+              auto_scaling_group_name: resource_id
+            ).flat_map(&:scaling_policies).flat_map(&:alarms).map(&:alarm_arn)
+            resource_client = cloudwatch_client
+          else
+            next
+          end
```

Comment:
Any chance you could make this more object oriented?   There's too much knowledge of lower level resources here I think.  There are already POROs in aws-ruby for resources like ASGs.  You could add `alarm.rb` and `event_rule.rb`.  Then you could make a 

```ruby
def resources(types*)
  Enumerator.new do |yielder|
    client.list_stack_resources(stack_name: name).each do |response|
      response.stack_resource_summaries.each do |stack_resource_summary|
        # assuming a `from_stack_resource_summary` class method on each PORO
        resource = types.first{ |type| type.from_stack_resource_summary(stack_resource_summary) }
        next if resource.nil?
        yielder << resource # something like this, not sure if this is the exact right way
      end
    end
  end
end
```

The resources would contain their own client method, which means we can get rid of the `cloudwatch_client` and friends above.

Then this block of code would start to look like:

```ruby
resources(Alarm, EventRule, AutoScalingGroup).each do |resource|
  resource_tags = resource.tags
  missing_tags = stack_tags - resource_tags

  next if missing_tags.empty?

  ...

  resource.tag(missing_tags)
end
```

Reply:
I'm not a fan of trying all the resources in `resource = types.first{ |type| type.from_stack_resource_summary(stack_resource_summary) }`, I feel like if anything this should be a hash somewhere...

Reply:
Sure, it can be a factory method somewhere.

Reply:
Maybe the hash does belong in the stack class because list_stack_resource returns some interesting types like `AWS::Events::Rule` which is not really what the resource is named elsewhere.

Or a cloudformation module/class.

Reply:
There could be a `ResourceFactory` with a `from_stack_resource_summary` class method?  Internally can be a switch statement or hash.

Reply:
Added ResourceFactory